src/components/Auth.js

```diff
@@ -0,0 +1,207 @@
+        const { user } = jwt.decode(authorizationToken);
+
+        await keycloak.loadUserInfo();
+        // eslint-disable-next-line camelcase
+        const username = keycloak.userInfo?.preferred_username || user?.name;
+
+        return {
+          authenticationToken: keycloak.token,
+          authorizationToken,
+          isAuthenticated: true,
+          isAdmin: Boolean(user.groups.find(group => group.name === dbRoles.admin)),
+          hasWriteAccess: Boolean(user.groups.find(group => [dbRoles.admin, dbRoles.regular].includes(group.name))),
+          user,
+          username,
+        };
+      }
+
+      await keycloak.logout();
+      return undefined;
+    },
+  );
+
+  const { authorizationToken } = data ?? {};
+
+  useEffect(() => {
+    const unregister = fetchIntercept.register({
+      request: (fetchUrl, fetchConfig) => {
+        if (fetchUrl.startsWith(window._env_.API_BASE_URL)) {
+          const newConfig = { ...fetchConfig };
+
+          if (!newConfig.headers) {
+            newConfig.headers = {};
+          }
+          newConfig.headers.Authorization = authorizationToken;
+          return [fetchUrl, newConfig];
+        }
+        return [fetchUrl, fetchConfig];
+      },
+    });
+    return unregister;
+  }, [authorizationToken]);
+
+  const auth = useMemo(() => ({
+    login: () => {
+      if (!isAuthenticating) {
+        logInOrOut(true);
+      }
+    },
+    logout: () => {
+      if (!isAuthenticating) {
+        logInOrOut();
+      }
+    },
+    isAuthenticating,
+    error,
+    ...data || {},
+  }), [data, isAuthenticating, logInOrOut, error]);
+
+  return (
+    <AuthContext.Provider
+      value={auth}
+    >
+      {children}
+    </AuthContext.Provider>
+  );
+};
+AuthProvider.propTypes = {
+  children: PropTypes.node.isRequired,
+};
+
+const Centered = ({ children }) => (
+  <div>
+    <Box alignItems="center" display="flex" flexDirection="column" paddingTop="40px" width="100%">{children}</Box>
+  </div>
+);
+Centered.propTypes = {
+  children: PropTypes.node.isRequired,
+};
+
+const AuthenticatedRoute = (props) => {
+  const { admin, component, path } = props;
+  const auth = useAuth();
+  const x = useRef();
+
+  useLayoutEffect(() => {
+    if (!auth.isAuthenticating && !auth.isAuthenticated && !x.current && !auth.error) {
+      auth.login();
+    }
+  }, [auth]);
+
+  if (auth.error) {
+    return (
+      <Route
+        path={path}
+      >
+        <Centered>
+          <Typography color="error" gutterBottom variant="h2">Error Authenticating</Typography>
+          <Typography paragraph>An Error occurred while authenticating. please logout and try again or contact your administrator if the problem persists</Typography>
+          <Typography paragraph>{auth.error?.message}</Typography>
+          <Button onClick={auth.logout}>logout</Button>
+        </Centered>
+      </Route>
+    );
+  }
+
+  if (!auth.isAuthenticated) {
```

Comment:
Should this check the isAuthorized rather than isAuthenticated? they shouldn't be able to see any data before they are authorized

Reply:
there is no `isAuthorized` only `isAuthenticated`. if they have both tokens `isAuthenticated` is `true`, otherwise `false`

Reply:
oic, those used to be separate! I missed that they were combined now